### PR TITLE
BIP 371: Taproot Fields for PSBT

### PR DESCRIPTION
**Pull request by achow101**
Specifies new field types and UTXO type semantics for Taproot PSBT fields.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Seems good to me. I don't think there is a need for separate field, `PSBT_IN_TAP_BIP32_DERIVATION` as we can stuff it in current fields. Current parser would fail, but anyway nothing can be expected to be done properly if the parser does not know `PSBT_IN_TAP_BIP32_DERIVATION`  either.

**Comment by NicolasDorier** [comment]
Need to mention some fields that need to go during finalize.
`PSBT_IN_TAP_INTERNAL_KEY `  and `PSBT_IN_TAP_KEY_SIG` need to be emptied when finalize is called.

**Comment by NicolasDorier** [comment]
isn't `PSBT_OUT_TAP_MERKLE_ROOT` needed?

File: bip-0174.mediawiki [hunk 1]
```diff
@@ -428,6 +428,72 @@ The currently defined per-input types are defined as follows:
...
+| The 64 or 65 byte Schnorr signature for key path spending a Taproot output.
+|
+|
+| 0, 2
+| bip-taproot-psbt
+|-
+| Taproot Script Spend Signature
+| <tt>PSBT_IN_TAP_SCRIPT_SIG = 0x14</tt>
+| <tt><xonlypubkey> <leafhash></tt>
+| The 32 byte X-only public key concatenated with the 32 byte hash of the leaf it is part of.
```
Comment (NicolasDorier) [inline bip-0174.mediawiki:445]: the internal key or the output key?

**Comment by achow101** [comment]
> isn't `PSBT_OUT_TAP_MERKLE_ROOT` needed?

I don't think so. I thought it might be, but could not see any situation where it would be needed but that the signer wouldn't already be able to get it from `PSBT_OUT_TAP_TREE`. These output fields are really for verifying that a change output is actually change. That would require checking that the scripts the signer cares about are in the taproot tree (hence `PSBT_OUT_TAP_TREE`), and then compute the merkle root from the tree, combine it with the internal key, and calculate the output key. If the merkle root were available, they still couldn't skip the steps to compute the merkle root from the tree to double check it as otherwise a merkle root could be provided that does not match the tree. So I don't see how providing the merkle root would necessarily help.

File: bip-0174.mediawiki (hunk 1 above)
Comment (achow101) [reply bip-0174.mediawiki:445]: Neither. This field is for any key that appears in a leaf script in the taproot tree.

**Comment by achow101** [comment]
> Need to mention some fields that need to go during finalize.
> `PSBT_IN_TAP_INTERNAL_KEY ` and `PSBT_IN_TAP_KEY_SIG` need to be emptied when finalize is called. (maybe there are others)

I believe all of the input fields can be removed after finalizing. I have mentioned this for each one.

File: bip-0174.mediawiki [hunk 2]
```diff
@@ -511,6 +577,50 @@ determine which outputs are change outputs and verify that the change is returni
...
+| The script for this leaf as would be provided in the witness stack.
+|
+|
+| 0, 2
+| bip-taproot-psbt
+|-
+| Taproot Key BIP 32 Derivation Path
+| <tt>PSBT_OUT_TAP_BIP32_DERIVATION = 0x07</tt>
+| <tt><xonlypubkey></tt>
+| The 32 byte X-only public key
```
Comment (NicolasDorier) [inline bip-0174.mediawiki:616]: output or internal?

Comment (NicolasDorier) [reply bip-0174.mediawiki:616]: answered https://github.com/bitcoin/bips/pull/1139/files#r670069639

**Comment by NicolasDorier** [comment]
For `PSBT_IN_TAP_BIP32_DERIVATION ` what do we put there for a key spend? 
I can't find where to put the keypath of the key spend.

File: bip-taproot-psbt.mediawiki [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+| The script for this leaf as would be provided in the witness stack followed by the single byte leaf version. Note that the leaves included in this field should be those that the signers of this input are expected to be able to sign for. Finalizers should remove this field after <tt>PSBT_IN_FINAL_SCRIPTWITNESS</tt> is constructed. Finalizers should remove this field after <tt>PSBT_IN_FINAL_SCRIPTWITNESS</tt> is constructed.
+|
+|
+| 0, 2
+|-
+| Taproot Key BIP 32 Derivation Path
+| <tt>PSBT_IN_TAP_BIP32_DERIVATION = 0x16</tt>
+| <tt><xonlypubkey></tt>
+| A 32 byte X-only public key involved in this input. It may be the internal key, or a key present in a leaf script.
+| <tt><hashes len> <leaf hash>* <32-bit uint> <32-bit uint>*</tt>
```
Comment (NicolasDorier) [inline bip-taproot-psbt.mediawiki:82]: for an internal key, I suppose `hashes len` is 0

Comment (NicolasDorier) [reply bip-taproot-psbt.mediawiki:82]: Why we don't have the master key fingerprint? The current hd path has it.

Comment (achow101) [reply bip-taproot-psbt.mediawiki:82]: Yes. I've written that in.

Comment (NicolasDorier) [reply bip-taproot-psbt.mediawiki]: For key spend, do you need to put the internal key with 0, or the output key with 0?

Comment (achow101) [reply bip-taproot-psbt.mediawiki]: The internal key.

The output key is basically irrelevant for these fields.

**Comment by NicolasDorier** [comment]
I created one, should be OK.
```
cHNidP8BAFUBAAAAAeDVt3US5q29vUNZcRX3LEdkDhgziyJ99ikJw2/8Dn7aAQAAAAD/////AZng9QUAAAAAGXapFHrYwgL4Q4eR3dKl1LYsqM39kgFhiKwAAAAAAAEBKwDh9QUAAAAAIlEg1iv7bdF2yzf4CpjhDnVXLsIfNrt5+1jrtX/uL1EgulwBGCAOu6VEp72Zgvt0Llb+yL7hR/3UAFTVdjZ8OO41pIs0WwEXIFMl9e2ZwmDOoNae5zPxtJO+bHS2kdFJeO3BVG6nFgeWARNACj/rUgWiFLb9ZWOdt03rW1sjiGDBiVKORiTD8CIA99SwCwkbA5/2TW/HAxKpSgISzX+B+WhQ/MHYSbete4IqBCEW1iv7bdF2yzf4CpjhDnVXLsIfNrt5+1jrtX/uL1Egulw5ASfq5OMcd3GONPFnf3KVRgqurQ8fvNI+EMf01mCCAHmxEjRWeFYAAIAAAACAAAAAgAAAAAAAAAAAAAA=
```
```json
{
  "fee": "0.00000103 BTC",
  "feeRate": "1 Sat/B",
  "tx": {
    "hash": "0ffc6c3faeef6228485290b60068803bd5d46f2c44d0d3aaaff833b3f313066e",
    "ver": 1,
    "vin_sz": 1,
    "vout_sz": 1,
    "lock_time": 0,
    "size": 85,
    "in": [
      {
        "prev_out": {
          "hash": "da7e0efc6fc30929f67d228b33180e64472cf715715943bdbdade61275b7d5e0",
          "n": 1
        },
        "scriptSig": ""
      }
    ],
    "out": [
      {
        "value": "0.99999897",
        "scriptPubKey": "OP_DUP OP_HASH160 7ad8c202f8438791ddd2a5d4b62ca8cdfd920161 OP_EQUALVERIFY OP_CHECKSIG"
      }
    ]
  },
  "inputs": [
    {
      "index": 0,
      "taproot_internal_key": "5325f5ed99c260cea0d69ee733f1b493be6c74b691d14978edc1546ea7160796",
      "taproot_merkle_root": "5b348ba435ee387c3676d55400d4fd47e1bec8fe562e74fb8299bda744a5bb0e",
      "taproot_key_signature": "0a3feb5205a214b6fd65639db74deb5b5b238860c189528e4624c3f02200f7d4b00b091b039ff64d6fc70312a94a0212cd7f81f96850fcc1d849b7ad7b822a04",
      "partial_signatures": {},
      "witness_utxo": {
        "value": "1.00000000",
        "scriptPubKey": "1 d62bfb6dd176cb37f80a98e10e75572ec21f36bb79fb58ebb57fee2f5120ba5c"
      },
      "bip32_derivs": [],
      "bip32_taproot_derivs": [
        {
          "pubkey": "d62bfb6dd176cb37f80a98e10e75572ec21f36bb79fb58ebb57fee2f5120ba5c",
          "master_fingerprint": "12345678",
          "path": "86'/0'/0'/0/0",
          "leaf_hashes": [
            "b179008260d6f4c7103ed2bc1f0fadae0a4695727f67f1348e71771ce3e4ea27"
          ]
        }
      ]
    }
  ],
  "outputs": [
    {
      "bip32_derivs": [],
      "bip32_taproot_derivs": []
    }
  ]
}
```

**Comment by achow101** [comment]
> I created one, should be OK.

Looks correct to me.

***

I've been working on a reference implementation and first draft of it is at https://github.com/achow101/bitcoin/tree/taproot-psbt. It still needs some testing, but once that is done, I'll list it in the BIP and add test vectors.

**Comment by luke-jr** [comment]
Assigned BIP number 371

**Pull request merged**